prov/verbs/src/fi_verbs.c

```diff
@@ -319,10 +319,16 @@ fi_ibv_send(struct fi_ibv_msg_ep *ep, struct ibv_send_wr *wr, void *context)
 
 	if (ep->scq) {
 		if (wr->send_flags & IBV_SEND_SIGNALED) {
+			fastlock_acquire(&ep->wre_lock);
 			wre = util_buf_alloc(ep->wre_pool);
-			if (OFI_UNLIKELY(!wre))
+			if (OFI_UNLIKELY(!wre)) {
+				fastlock_release(&ep->wre_lock);
 				return -FI_EAGAIN;
+			}
 			memset(wre, 0, sizeof(*wre));
+			dlist_insert_tail(&wre->entry, &ep->wre_list);
+			fastlock_release(&ep->wre_lock);
```

Comment:
It seems on send error FI_IBV_RELEASE_WRE doesn't lock when releasing the entry.

Reply:
Good catch. Has been updated according your comment